app/services/db/monthly_sp_auth_count/unique_monthly_auth_counts_by_partner.rb

```diff
@@ -0,0 +1,144 @@
+        date_range = start_date...end_date
+
+        return [] if !date_range || issuers.blank?
+
+        # Query a month at a time, to keep query time/result size fairly reasonable
+        # The results are rows with [user_id, ial, auth_count, year_month]
+        months = Reports::MonthHelper.months(date_range)
+        queries = build_queries(issuers: issuers, months: months)
+
+        ial_to_year_month_to_users = Hash.new do |ial_h, ial_k|
+          ial_h[ial_k] = Hash.new { |ym_h, ym_k| ym_h[ym_k] = Multiset.new }
+        end
+
+        queries.each do |query|
+          temp_copy = ial_to_year_month_to_users.deep_dup
+
+          with_retries(
+            max_tries: 3,
+            rescue: [
+              ActiveRecord::SerializationFailure,
+              PG::ConnectionBad,
+              PG::TRSerializationFailure,
+              PG::UnableToSend,
+            ],
+            handler: proc do
+              ial_to_year_month_to_users = temp_copy
+              ActiveRecord::Base.connection.reconnect!
+            end,
+          ) do
+            Reports::BaseReport.transaction_with_timeout do
+              ActiveRecord::Base.connection.execute(query).each do |row|
+                user_id = row['user_id']
+                year_month = row['year_month']
+                ial = row['ial']
+                if row['min_returned_at'].nil? 
+                  next
+                elsif row['max_profile_verified_at'].nil?
+                  profile_age = -1
+                else
+                  profile_age = (Time.zone.parse(row['min_returned_at'].to_s).year - Time.zone.parse(row['max_profile_verified_at'].to_s).year)
+                end
+                ial_to_year_month_to_users[ial][year_month].add([user_id, profile_age])
+              end
+            end
+          end
+        end
+
+        rows = []
+
+        ial_to_year_month_to_users.each do |ial, year_month_to_users|
+          prev_seen_users = Set.new
+
+          year_months = year_month_to_users.keys.sort
+
+          year_months.each do |year_month|
+            year_month_users = year_month_to_users[year_month]
+
+            auth_count = year_month_users.count
+            unique_users = year_month_users.uniq.to_set
+
+            new_unique_users = unique_users - prev_seen_users
+            new_unique_users_year1 = new_unique_users.select { |user_id, profile_age| profile_age == 0 }
+            new_unique_users_year2 = new_unique_users.select { |user_id, profile_age| profile_age == 1 }
+            new_unique_users_year3 = new_unique_users.select { |user_id, profile_age| profile_age == 2 }
+            new_unique_users_year4 = new_unique_users.select { |user_id, profile_age| profile_age == 3 }
+            new_unique_users_year5 = new_unique_users.select { |user_id, profile_age| profile_age == 4 }
+            new_unique_users_year_greater_than_5 = new_unique_users.select { |user_id, profile_age| profile_age > 4 }
```

Comment:
we could probably use the `Enumerable#count` method to save a step (and not have to keep a potentially large array in memory)
```suggestion
            new_unique_users_year1 = new_unique_users.count { |_user_id, profile_age| profile_age == 0 }
            new_unique_users_year2 = new_unique_users.count { |_user_id, profile_age| profile_age == 1 }
            new_unique_users_year3 = new_unique_users.count { |_user_id, profile_age| profile_age == 2 }
            new_unique_users_year4 = new_unique_users.count { |_user_id, profile_age| profile_age == 3 }
            new_unique_users_year5 = new_unique_users.count { |_user_id, profile_age| profile_age == 4 }
            new_unique_users_year_greater_than_5 = new_unique_users.select { |_user_id, profile_age| profile_age > 4 }
```

Reply:
Thanks for the feedback we will implement and test